Correctly handle null characters if the length is known

Demindiro opened the pull request:
Closes #44478

Demindiro committed:
Ignore null character if p_len is specified

Closes #44478

bruvzg reviewed (commented):
Seems fine. I do not see anything wrong with supporting NULs in the middle of the string.

Note: some other string functions stop on 0x0000 instead of using length as well: `get_slicec`, `hex_to_int`, `bin_to_int` for example.

bruvzg commented:
Unrelated thought: maybe we want to support overlong NUL character representation used by Java (`0xC0 0x80`) too. Right now all overlong chars are rejected, but I guess UTF-8 strings received from Java on Android can have it.

Demindiro committed:
Account for string length in {hex,bin}_to_int

Demindiro committed:
Use *this instead of this->ptr() in copy constructor

Demindiro committed:
Pass string length to utf8() in percent_decode()

Demindiro committed:
Use int instead of size_t

This is to make CI not fail (I wonder why it compiles fine on my machine
though...)

Also, why does length() return an int anyways? Wouldn't size_t make more
sense?

Demindiro committed:
Please clang-format by removing whitespace

Sheesh

Demindiro committed:
Make get_slicec account for actual string length

Demindiro commented:
I think this covers everything in `ustring.cpp`. Grepping for `ptr()` gives me:

```
$ rg 'ptr\(\)' core/string/ustring.cpp 
1023:		if (i == length() || ptr()[i] == p_splitter) {
2102:	int64_t sign = ptr()[i] == '-' ? -1 : 1;
2109:		if (ptr()[i] != '0' || ptr()[i + 1] != 'x') {
2118:		char32_t c = LOWERCASE(ptr()[i]);
2144:	int64_t sign = ptr()[i] == '-' ? -1 : 1;
2151:		if (ptr()[i] != '0' || ptr()[i + 1] != 'b') {
2160:		char32_t c = LOWERCASE(ptr()[i]);
2679:	CryptoCore::md5((unsigned char *)cs.ptr(), cs.length(), hash);
2686:	CryptoCore::sha1((unsigned char *)cs.ptr(), cs.length(), hash);
2693:	CryptoCore::sha256((unsigned char *)cs.ptr(), cs.length(), hash);
2700:	CryptoCore::md5((unsigned char *)cs.ptr(), cs.length(), hash);
2713:	CryptoCore::sha1((unsigned char *)cs.ptr(), cs.length(), hash);
2727:	CryptoCore::sha256((unsigned char *)cs.ptr(), cs.length(), hash);
4388:	return String::utf8(pe.ptr(), pe.length());
4763:	copymem(w, charstr.ptr(), len);
4779:	copymem(w, charstr.ptr(), len);
4795:	copymem(w, (const void *)charstr.ptr(), len);
4810:	copymem(w, (const void *)s->ptr(), len);
```

I don't think there is an alternative to `ptr()` that I overlooked either.

akien-mga commented:
How is all this refactoring around `ptr()` related to th original issue? This makes the PR much harder to review and much more prone to create regressions in a core piece of the API.

Demindiro commented:
> How is all this refactoring around `ptr()` related to the original issue? This makes the PR much harder to review and much more prone to create regressions in a core piece of the API.
> 
> I'm not saying the change is wrong but it should be in a separate PR if it's not related to the handling of null characters.

What exactly do you mean? For functions like `bin_to_int` I believe using `ptr()[index]` is the best way to avoid things like `p < ptr() + length()`. Removing `this->` did end up affecting more than I anticipated, so I suppose I should revert that.

bruvzg commented on core/string/ustring.cpp:
```diff
@@ -1016,22 +1016,20 @@ String String::get_slicec(char32_t p_splitter, int p_slice) const {
 	int count = 0;
 	while (true) {
-		if (c[i] == 0 || c[i] == p_splitter) {
+		if (i == length() || this->ptr()[i] == p_splitter) {
```
In debug builds, calling `this->ptr()[]` on each step is 3...5 times slower than using `const char32_t *c = this->ptr();` once!

akien-mga commented:
We discussed this on chat with @bruvzg and @Faless and tend to err on the side of not allowing null characters in String:

> [bruvzg] Are there legitimate uses for it? `NUL` is a valid Unicode character, but it's probably better to just document that strings are null-terminated and can't have `NUL` character. If it's desired, we should add support to more places than this PR adds. I plan (probably this week) to add support for decoding some non-standard variants of UTF-8 / UTF-16 encoding. Windows file paths can have invalid unpaired UTF-16 surrogates, and it's possible to get CESU-8/WTF-8 encoded stuff from external sources, so better to support it (with a warning) rather than rejecting such strings. As well as add more readable string decoding errors (instead of generic "Unicode parsing error"). So I can add `NUL` support at the same time.
> [Akien] No idea. The original bug report is godot#44478 maybe fales can chime in.
> [Akien] If there's no clear need for it and it adds significant complexity, then I'm more for the "Do Nothing" approach of our guidelines 🙂
> [fales] not sure, the streempeerbuffer code seems to just rely on `String` doing the parsing (and encodes the length it receives from it). While `NUL` appears to be valid UTF-8 any string containing the `NUL` character **will not be compatible with C string operations** (which treats `NUL` as the terminator). I have no clue why they decided to include it in UTF-8
> [fales] So, any usage of `str.utf8().get_data()` passed to a C `str*` function will be cut short at the first 0x0
> [Akien] Then I think that we should probably not support it to stay compatible with C strings, which is quite important for GDExtension. And make it explicit in the docs.
```

akien-mga commented:
Some documentation suggestion:
```diff
diff --git a/doc/classes/StreamPeer.xml b/doc/classes/StreamPeer.xml
index bd69867001..6e4b13a147 100644
--- a/doc/classes/StreamPeer.xml
+++ b/doc/classes/StreamPeer.xml
@@ -71,6 +71,7 @@
 			<argument index="0" name="bytes" type="int" default="-1" />
 			<description>
 				Gets an ASCII string with byte-length [code]bytes[/code] from the stream. If [code]bytes[/code] is negative (default) the length will be read from the stream using the reverse process of [method put_string].
+				[b]Note:[/b] If the stream includes a [code]NUL[/code] character (byte 0) before the given [code]bytes[/code] length, the decoded string will be truncated at that first [code]NUL[/code] character.
 			</description>
 		</method>
 		<method name="get_u16">
@@ -102,6 +103,7 @@
 			<argument index="0" name="bytes" type="int" default="-1" />
 			<description>
 				Gets an UTF-8 string with byte-length [code]bytes[/code] from the stream (this decodes the string sent as UTF-8). If [code]bytes[/code] is negative (default) the length will be read from the stream using the reverse process of [method put_utf8_string].
+				[b]Note:[/b] If the stream includes a [code]NUL[/code] character (byte 0) before the given [code]bytes[/code] length, the decoded string will be truncated at that first [code]NUL[/code] character.
 			</description>
 		</method>
 		<method name="get_var">
@@ -182,6 +184,7 @@
 				PutData("Hello World".ToAscii());
 				[/csharp]
 				[/codeblocks]
+				[b]Note:[/b] If [code]value[/code] contains a [code]NUL[/code] character before the end of the string, it will be truncated.
 			</description>
 		</method>
 		<method name="put_u16">
@@ -226,6 +229,7 @@
 				PutData("Hello World".ToUTF8());
 				[/csharp]
 				[/codeblocks]
+				[b]Note:[/b] If [code]value[/code] contains a [code]NUL[/code] character before the end of the string, it will be truncated.
 			</description>
 		</method>
 		<method name="put_var">
```

But I think `String` should also have more details in the description about whether it's null terminated and how it handles null characters mid string.

akien-mga commented:
Superseded by #62735.
